pkg/suggestion/v1beta1/skopt/base_service.py

```diff
@@ -112,7 +112,9 @@ def getSuggestions(self, trials, current_request_number):
                 logger.info("List of recorded Trials names: {}\n".format(self.recorded_trials_names))
 
         else:
-            logger.info("Succeeded Trials didn't change: {}\n".format(self.succeeded_trials))
+            logger.error("Succeeded Trials didn't change: {}\n".format(self.succeeded_trials))
+            logger.error("No new suggestions could be generated, return early..\n")
+            return []
```

Comment:
@shaowei-su What if Trial was failed for some reason (e.g. networking) and we ask for the new Trials ?
In that case `succeeded_trials` list will stay the same.

Reply:
That's a great question @andreyvelich , I can think of two possible routes:

1. Add more conditional branches in the suggestions service, e.g is the service invoked as a result of failed trials? if so, suggestion service will return duplicated suggestions as before this PR change;

2. Retry for trials should be kept as a logic within trial templates themself i.e Job, TFJob, MPIJob etc.. HPO search experiment only make progress when there are succeeded (including early stopped ones) trials complete. In the case of multiple concurrent trials, suggestion service will not return new trials until next succeeded trial's ready.

Reply:
> Add more conditional branches in the suggestions service, e.g is the service invoked as a result of failed trials?

But do we need conditional branch to the suggestion service ?
If `succeeded_trials` = `Succeeded Trials` + `Early Stopped Trials`, then if Trial was failed the Suggestion service will just ask for the new trials (e.g. with duplicated HPs) using `current_request_number`. 

> Retry for trials should be kept as a logic within trial templates themself i.e Job, TFJob, MPIJob etc.. 

That's valid assumption. In that case, what should be the nature of [MaxFailedTrialCount](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1beta1/experiment_types.go#L51) API ?

cc @johnugeorge @tenzen-y

Reply:
> But do we need conditional branch to the suggestion service ?

The reason is that suggestion service can't tell whether the `current_request_number` is requested as a result of failed trails (in this case service should return duplicated HPs), or as a result of delay in trial status/observations updates (in this case service should wait until trials are fully updated).

The latter can be reproduced by this [example](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/early-stopping/median-stop.yaml#L17) and change the algo to `bayesianoptimization`.

Reply:
@shaowei-su Should we maintain this logic in our skopt suggestion ? E.g. maintain list of created Trials in Suggestion service (similar to our [Chocolate Service](https://github.com/kubeflow/katib/blob/1613191656368e0d9d84a7de1615153584814a99/pkg/suggestion/v1beta1/chocolate/base_service.py#L54)) ?  Then, if Suggestion was invoked, but no new Trials were Succeeded we just return HPs from that list.
WDYT ?

Reply:
@andreyvelich I checked the logic from Chocolate Service and not quite sure if that can solve the issue here, e.g when trial early stopped without observation updated in time, if suggestion service consider there is no new succeeded trials, and returning HPs from created list will generate duplicated trials.

Another solution is to root cause why suggestion service is called with `current_request_number > 0` before trial observation is fully updated for early stopping. In theory, [trial_controller](https://github.com/kubeflow/katib/blob/v0.14.0/pkg/controller.v1beta1/trial/trial_controller.go#L255) should guarantee that metric is fully up to date.

Reply:
hmm I think the reason might be that for early stopped trials, the status is updated by a separate service call (https://github.com/kubeflow/katib/blob/master/pkg/earlystopping/v1beta1/medianstop/service.py#L184) in async, whereas for succeeded trials update metrics and change status happen in the same trial controller loop (https://github.com/kubeflow/katib/blob/v0.14.0/pkg/controller.v1beta1/trial/trial_controller.go#L264).

To fix this, we could try to add the early stopped but non-observation ready trials into the active trials count as well. wdyt?

Reply:
> when trial early stopped without observation updated in time

Should we avoid sending Early Stopped Trials without observation to the Suggestion service ? e.g. we don't send Trials with [Metrics Unavailable condition](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go#L343-L344). I think, Trials logically shouldn't be called "EarlyStopped" until Observation is ready for them. WDYT @shaowei-su @johnugeorge @tenzen-y ?

I know it is hacky that we set EarlyStopping trial condition from the Early Stopping service and only after that we add Observation Trial results. We might want to reconsider some early stopping design when we introduce push-based Metrics Collector: https://github.com/kubeflow/katib/issues/577

Reply:
> To fix this, we could try to add the early stopped but non-observation ready trials into the active trials [count](https://github.com/kubeflow/katib/blob/v0.14.0/pkg/controller.v1beta1/experiment/experiment_controller.go#L293) as well. wdyt?

Since Suggestion and Trial controller works in async manner, I think we should introduce another check to Suggestion Client as I mentioned above.

Reply:
Actually those trials are already [filtered out](https://github.com/kubeflow/katib/blob/master/pkg/suggestion/v1beta1/internal/trial.py#L61) on the suggestion service side.

What's missing here is the correct [currentRequestNum](https://github.com/kubeflow/katib/blob/master/pkg/controller.v1beta1/suggestion/suggestionclient/suggestionclient.go#L88), which should not be incremented before the early stopping metrics are fully updated.

Reply:
@shaowei-su But is that correct approach to filter them on the Suggestion side instead of Katib Controller side ?

Some of the suggestion services might not use `internal` implementation and just forget about it.
I think the Katib controller should just invoke suggestion service with the correct statuses for the Trials. 
E.g. [these Trial statuses](https://github.com/kubeflow/katib/blob/master/pkg/apis/manager/v1beta1/api.proto#L232-L242).
Which means if EarlyStopped Trial doesn't have observation, Controller should not send it to Suggestion Service.
The main purpose of the Suggestion service is just produce new HPs based on Trial results.

I doubt that Suggestion service should do any additional work with Trials from Katib controller.

What do other think about it ? @johnugeorge @gaocegege @tenzen-y